### PR TITLE
Optimize ecdsa

### PR DESCRIPTION
**Pull request by Megamouse**
Improves #14161

```
Old: Package installation took 212.59543 seconds
New: Package installation took  ? seconds
```

### CODE REVIEW

File: rpcs3/Crypto/ec.cpp [hunk 1]
```diff
@@ -204,89 +190,80 @@ static thread_local point ec_G{}; // mon
...
+static inline bool elt_is_zero(const u8* d)
 {
-	u32 i;
-
-	for (i = 0; i < 20; i++)
-		if (d[i] != 0)
-			return 0;
-
-	return 1;
+	static const u8 zero[20]{};
```
Comment (elad335) [inline rpcs3/Crypto/ec.cpp:205]: constexpr

File: rpcs3/Crypto/ec.cpp [hunk 2]
```diff
@@ -103,17 +97,14 @@ static constexpr u8 inv256[0x80] = {
...
-	u32 i;
-
 	const u8 z = -(d[n - 1] + a[n - 1] * b) * inv256[N[n - 1] / 2];
 
-	dig = d[n - 1] + a[n - 1] * b + N[n - 1] * z;
+	u32 dig = d[n - 1] + a[n - 1] * b + N[n - 1] * z;
 	dig >>= 8;
 
-	for (i = n - 2; i < n; i--)
+	for (u32 i = n; i >= 2; i--)
```
Comment (elad335) [inline rpcs3/Crypto/ec.cpp:107]: Is it actually the same? i is used for indexing.

Comment (Megamouse) [reply rpcs3/Crypto/ec.cpp:107]: It looks like it's relying on the overflow to exit the loop, which is an ill defined loop.

File: rpcs3/Crypto/ec.cpp [hunk 3]
```diff
@@ -30,55 +30,49 @@ static int bn_compare(u8* a, u8* b, u32 n)
...
+static inline u8 bn_add_1(u8* d, const u8* a, const u8* b, u32 n)
 {
-	u32 i;
-	u32 dig;
-	u8 c;
+	u8 c = 0;
 
-	c = 0;
-	for (i = n - 1; i < n; i--)
+	for (u32 i = n; i > 0; i--)
```
Comment (elad335) [inline rpcs3/Crypto/ec.cpp:37]: I think either i != umax or check i == 0 after tge operation (before i--)

Comment (elad335) [reply rpcs3/Crypto/ec.cpp:37]: i == 0 after operation would allow to not rely on overflow.

Comment (Megamouse) [reply rpcs3/Crypto/ec.cpp:37]: I could cast to s64 which would probably do the same.
But what you suggest is also ill-defined.

Comment (elad335) [reply rpcs3/Crypto/ec.cpp:37]: It's not, for (u32 i;; i--) { if (i == 0) break; }

Comment (Megamouse) [reply rpcs3/Crypto/ec.cpp]: Turns out the whole ecdsa internally only uses n = 20 and n = 21.
So I made them template functions and simply use s32.

**Comment by Megamouse** [comment]
something is still off though

File: rpcs3/Crypto/ec.cpp [hunk 4]
```diff
@@ -5,17 +5,8 @@
...
-}
-
-static void bn_copy(u8* d, u8* a, u32 n)
-{
-	memcpy(d, a, n);
-}
-
-static int bn_compare(u8* a, u8* b, u32 n)
+template <s32 n>
+static inline int bn_compare(const u8* a, const u8* b)
```
Comment (elad335) [inline rpcs3/Crypto/ec.cpp:9]: isnt this just memcmp? you clamp the result to +1 and -1 for completeness.

**Comment by Megamouse** [comment]
Turns out the optimization is fruitless since it needs the point copies after all. -> same time as before.

**Comment by Megamouse** [comment]
I made it run async for each entry.
Now it only takes 20 seconds to install Demon's souls instead of 200.

**Comment by Megamouse** [comment]
turns out I was basically sidestepping the verification due to thread locals.

**Pull request merged**